src/main/java/com/epam/aidial/core/upstream/UpstreamRoute.java

```diff
@@ -6,25 +6,26 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicLong;
 
 @RequiredArgsConstructor
 public class UpstreamRoute implements Iterator<Upstream> {
 
     private final List<Upstream> upstreams;
-    private final AtomicLong counter;
     private final int offset;
+    private final int maxAttempts;
 
-    private Upstream current;
-    private int count;
+    private Upstream upstream;
+    private int attempts;
+    private int next;
+    private int prev;
 
     public int attempts() {
-        return count;
+        return attempts;
     }
 
     @Override
     public boolean hasNext() {
-        return count < upstreams.size();
+        return next < upstreams.size() && attempts < maxAttempts;
```

Comment:
```
return count < upstreams.size() && attempts[(count + offset) % upstreams.size()] < maxAttempts
```

Reply:
What you are trying to solve?

Reply:
track the number of retries per upstream. But since we want to do that I discard the comment.